File: src/php/core/src/Style/ComponentVariant.php
<?php

declare(strict_types=1);

namespace DiceBear\Style;

/**
 * Read-only view over an entry in a component's `variants` block.
 */
class ComponentVariant
{
    /** @var list<Element>|null */
    private ?array $elements = null;

    /**
     * @param array<string, mixed> $data
     */
    public function __construct(private readonly array $data) {}

    /**
     * Returns the variant's elements, lazily wrapped as {@see Element}
     * instances on first access.
     *
     * @return list<Element>
     */
    public function elements(): array
    {
        return $this->elements ??= array_map(
            fn(array $el) => new Element($el),
            $this->data['elements'],
        );
    }

    /**
     * Returns the weighted-pick weight for this variant, defaulting to `1`.
     */
    public function weight(): int|float
    {
        return $this->data['weight'] ?? 1;
    }
}

File: src/rust/core/src/style/component_variant.rs
//! Read-only view over an entry in a component's `variants` block.

use serde::Deserialize;

use super::element::Element;

#[derive(Deserialize)]
pub(crate) struct ComponentVariant {
    elements: Vec<Element>,
    weight: Option<f64>,
}

impl ComponentVariant {
    pub(crate) fn elements(&self) -> &[Element] {
        &self.elements
    }

    /// The weighted-pick weight, defaulting to `1`.
    pub(crate) fn weight(&self) -> f64 {
        self.weight.unwrap_or(1.0)
    }
}
